tests/testsuite/registry_auth.rs

```diff
@@ -0,0 +1,247 @@
+    RegistryBuilder::new()
+        .alternative(true)
+        .auth_required()
+        .add_tokens(false)
+        .http(server.url())
+        .build();
+
+    let p = make_project();
+    cargo(&p, "build")
+        .with_status(101)
+        .with_stderr(
+            "\
+[UPDATING] `alternative` index
+[ERROR] failed to get `bar` as a dependency of package `foo v0.0.1 ([..])`
+
+Caused by:
+  no token found for `alternative`, please run `cargo login --registry alternative`",
+        )
+        .run();
+}
+
+#[cargo_test]
+fn incorrect_token() {
+    let server = serve_registry(alt_registry_path(), Some("invalid".to_string()));
+    RegistryBuilder::new()
+        .alternative(true)
+        .auth_required()
+        .http(server.url())
+        .build();
+
+    let p = make_project();
+    cargo(&p, "build")
+        .with_status(101)
+        .with_stderr(
+            "\
+[UPDATING] `alternative` index
+[ERROR] failed to get `bar` as a dependency of package `foo v0.0.1 ([..])`
+
+Caused by:
+  token rejected for `alternative`, please run `cargo login --registry alternative`
+
+Caused by:
+  remote server said: Unauthorized message from server.",
+        )
+        .run();
+}
+
+#[cargo_test]
+fn login() {
+    let server = serve_registry(alt_registry_path(), Some("sekrit".to_string()));
+    RegistryBuilder::new()
+        .alternative(true)
+        .add_tokens(false)
+        .auth_required()
+        .http(server.url())
+        .build();
+
+    let p = make_project();
+    cargo(&p, "login --registry alternative")
+        .with_stdout("please paste the token found on https://test-registry-login/me below")
+        .with_stdin("sekrit")
+        .run();
+}
+
+#[cargo_test]
+fn login_existing_token() {
+    let server = serve_registry(alt_registry_path(), Some("sekrit".to_string()));
+    RegistryBuilder::new()
+        .alternative(true)
+        .add_tokens(true)
+        .auth_required()
+        .http(server.url())
+        .build();
+
+    let p = make_project();
+    cargo(&p, "login --registry alternative")
+        .with_stdout("please paste the token found on https://test-registry-login/me below")
+        .with_stdin("sekrit")
+        .run();
+}
+
+#[cargo_test]
+fn login_server_unavailable() {
+    // Set up a registry with an invalid url.
+    RegistryBuilder::new()
+        .alternative(true)
+        .add_tokens(false)
+        .auth_required()
+        .http(Url::parse("sparse+http://127.0.0.1:0/").unwrap())
+        .build();
+
+    let p = make_project();
+    cargo(&p, "login --registry alternative")
+        .with_stdout("please paste the token for alternative below")
+        .with_stdin("sekrit")
+        .run();
+}
+
+#[cargo_test]
+fn duplicate_index() {
+    RegistryBuilder::new()
+        .alternative(true)
+        .add_tokens(false)
+        .auth_required()
+        .build();
+    let p = make_project();
+
+    // Two alternative registries with the same index.
+    cargo(&p, "build")
```

Comment:
I think there's room to have a discussion here about whether this error should occur when the configuration is constructed, or when an ambiguous configuration is read.
- The advantage to construction time (Like this PR imps), is that it is clear to the user what is causing the problem.
- The advantage of read time, is that in most contexts this is not ambiguous. It only becomes ambiguous if you have multiple configured registries on the same URL and a registry depends on something else from that URL.

We can always loosen it in the future. So this is probably the right call. But I do want to bring up the issue.

Reply:
We could also be able to attempt to disambiguate by excluding registries that don't have tokens configured.

As you mentioned, I did it this way because it seemed easier to explain to the user.